# Seminario sesión 6. Análisis de datos usando Pandas

En este notebook vamos a aplicar lo visto en el notebook explicatorio a un nuevo fichero csv. Para cada ejercicio hay una celda de código vacía que tendrás que completar con la instrucción correspondiente. Puedes añadir nuevas celdas de código si lo consideras necesario.   

Al finalizar, recuerda guardar los cambios mediante la opción *Save a copy in GitHub...* del menú *File*.

## Ejercicios

Comenzamos descargando el dataset de los supervivientes del Titanic.

In [5]:
!wget https://raw.githubusercontent.com/IA1819/Datasets/master/titanic_train.csv -O titanic_train.csv

--2023-11-10 16:06:25--  https://raw.githubusercontent.com/IA1819/Datasets/master/titanic_train.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 60302 (59K) [text/plain]
Saving to: ‘titanic_train.csv’

titanic_train.csv   100%[===================>]  58.89K  --.-KB/s    in 0.01s   

2023-11-10 16:06:26 (4.53 MB/s) - ‘titanic_train.csv’ saved [60302/60302]



Carga las librerías pandas y numpy.

In [6]:
import pandas as pd
import numpy as np

Carga el dataset que hemos descargado (llamado `titanic_train.csv`) en la variable `df`.

In [7]:
df=pd.read_csv("titanic_train.csv")

Muestra las primeras instancias de dicho dataset.

In [8]:
df # Muestra las cinco primeras filas y las cinco últimas (mi opción inicial)
df.head() #Muestra las cinco primeras filas


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

Muestra los tipos de los distintos descriptores del dataset.

In [9]:
"""Hago la prueba con dos columnas para ver si el tipo es el mismo para las dos, que lo es. En mis búsquedas veo que en
"series.Series", "series" es la instancia dentro de "Series", que es una estructura de datos unidimensional
que se utiliza para representar una columna o una sola variable en un DataFrame de Pandas."""
type1=type(df["PassengerId"])
type2=type(df["Survived"])
type_total=type(df)
print(type1)
print(type2)
print(type_total)
#Tras ver el cuaderno de apoyo, utilizo esta otra orden
print(df.info())

<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
None


Muestra los pasajeros que embarcaron en Cherbourg (Embarked=='C') y pagaron más de 200 libras por su ticket (Fare > 200).

In [10]:
pasajeros_seleccionados = df[(df["Embarked"] == "C") & (df["Fare"] >200)]
#Me cuesta la estructura: que haya dos "df" seguidos 'df[(df["Embarked"])].
print(pasajeros_seleccionados)

     PassengerId  Survived  Pclass  \
118          119         0       1   
258          259         1       1   
299          300         1       1   
311          312         1       1   
377          378         0       1   
380          381         1       1   
557          558         0       1   
679          680         1       1   
700          701         1       1   
716          717         1       1   
737          738         1       1   
742          743         1       1   

                                                  Name     Sex   Age  SibSp  \
118                           Baxter, Mr. Quigg Edmond    male  24.0      0   
258                                   Ward, Miss. Anna  female  35.0      0   
299    Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0      0   
311                         Ryerson, Miss. Emily Borie  female  18.0      2   
377                          Widener, Mr. Harry Elkins    male  27.0      0   
380                            

Vamos a crear un nuevo descriptor con la edad de los pasajeros, para lo que definimos la siguiente función.

In [11]:
def age_category(age):
    '''
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    '''
    if age < 30:
        return 1
    elif age < 55:
        return 2
    elif age >= 55:
        return 3

A continuación añadimos el nuevo descriptor usando el método `apply`.

In [12]:
df['Age_category'] = df['Age'].apply(age_category)

Responde a las siguientes preguntas. Para ello, incluye el código necesario y edita la celda que contiene el texto **Respuesta.**

¿Cuál era la proporción entre hombres y mujeres que iba a bordo? (Pista. Tendrás que utilizar el descriptor `Sex`)

In [13]:
proporcion_hombres= len(df[df["Sex"] == "male"]) / len(df["Sex"]) *100
#Inicialmente había puesto la función "len" en sitios equivocados:
#"proporcion_hombres= df[len(df["Sex"] == "male")] / (len(df["Sex"]))". Me cuesta el uso de paréntesis y corchetes.
a = int(proporcion_hombres)
print(f"La proporción de hombres a bordo era del: ", a, " %")
print(f"La proporción de hombres a bordo era del: ", proporcion_hombres, " %")
#Tengo que ver cómo limitar el número de decimales en el float a dos

La proporción de hombres a bordo era del:  64  %
La proporción de hombres a bordo era del:  64.75869809203144  %


La proporción de hombres a bordo era de casi el 65% (64,75%).

Utiliza una tabla de contingencia para averiguar cuántos hombres iban en segunda clase. (Pista. La clase en la que iban los pasajeros corresponde al descriptor `Pclass`)

In [17]:
pd.crosstab(df['Sex'] == "male", df['Pclass'] == 2)

Pclass  False  True 
Sex                 
False     238     76
True      469    108

108 hombres iban en segunda clase.

¿Cuál era la media y la desviación típica del precio de los tickets (descriptor `Fare`)?

In [21]:
#media
a=df['Fare'].mean()
print(f"La media de del precio de los tickets era: ",a)
#desviación típica. Solo veo la posibilidad de a través del método describe
df.describe()
'''Veo que a través del método describe podía haber obtenido los dos parámetros;
imagino que habrá un método para calcular solo la DT, al igual que la media
con mean'''

La media de del precio de los tickets era:  32.204207968574636


PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  Age_category  
count  891.000000  891.000000    714.000000  
mean     0.381594   32.204208      1.521008  
std      0.806057   49.693429      0.606400  
min      0.000000    0.000000      1.000000  
25%      0.000000    7.910400      1.000000  
50%      0.000000   14.454200      1.000000  
75%      0.000000   31.000000      2.000000  
max      6.000000  512.329200      3.000000

La media del precio de los tickets era de 32,20(€?) y la DT, de 49,69€.

¿Es verdad que la media de edad de los supervivientes es inferior a la de los pasajeros que fallecieron? (Pista. Puedes agrupar por el atributo `Survived`).

In [26]:
columns_to_show = ["Survived", "Age"]
df.groupby("Survived")[columns_to_show].describe(percentiles=[])

Survived                             Age                              \
            count mean  std  min  50%  max  count       mean        std   min   
Survived                                                                        
0           549.0  0.0  0.0  0.0  0.0  0.0  424.0  30.626179  14.172110  1.00   
1           342.0  1.0  0.0  1.0  1.0  1.0  290.0  28.343690  14.950952  0.42   

                      
           50%   max  
Survived              
0         28.0  74.0  
1         28.0  80.0

In [27]:
#Otra forma (tras búsquedas por internet)
media_edad_por_estado = df.groupby("Survived")["Age"].mean()
print(media_edad_por_estado)

Survived
0    30.626179
1    28.343690
Name: Age, dtype: float64


La media de edad de los 342 supervivientes fue de 28,34 años, mientras que la de los 549 pasajeros que fallecieron era de 30,62 años, con lo que sí, la media de los primeros era inferior a la de los segundos.

¿Es verdad que hubo más pasajeros menores de 30 años que sobrevivieron que aquellos mayores de 55?

In [29]:
df.groupby("Survived")[len("Age") < 30]

KeyError: ignored

**Respuesta**.

¿Es verdad que sobrevivieron más mujeres que hombres?

**Respuesta**.

## Entrega

Recuerda guardar el fichero con tu solución mediante la opción File -> Save a copy in GitHub.